crates/runtime/src/datafusion.rs

```diff
@@ -247,6 +260,58 @@ impl DataFusion {
         None
     }
 
+    pub fn set_cache_provider(&mut self, cache_provider: Option<QueryResultCacheProvider>) {
+        self.cache_provider = cache_provider;
+    }
+
+    pub async fn query_with_cache(&self, sql: &str) -> Result<Arc<Vec<RecordBatch>>> {
+        let session = self.ctx.state();
+        let plan = session
+            .create_logical_plan(sql)
+            .await
+            .context(UnableToExecuteQuerySnafu)?;
+
+        if let Some(cache_provider) = &self.cache_provider {
+            if let Some(cached_result) = cache_provider
+                .get(&plan)
+                .await
+                .context(FailedToAccessCacheSnafu)?
+            {
+                return Ok(cached_result);
+            }
+        }
+
+        let restricted_sql_options = SQLOptions::new()
+            .with_allow_ddl(false)
+            .with_allow_dml(false)
+            .with_allow_statements(false);
+
+        restricted_sql_options
+            .verify_plan(&plan)
+            .context(UnableToExecuteQuerySnafu)?;
+
+        let plan_copy = plan.clone();
```

Comment:
We can get rid of copy here by switching from using LogicalPlan for key to pre-calculated hash / u64